th-abstraction-0.5.0.0

Stackage is upgrading to th-abstraction-0.5.0.0, but recursion-schemes requires `< 0.5`. We must thus upgrade or we will be dropped from stackage.

From https://github.com/commercialhaskell/stackage/issues/6897 (thanks for the heads up!)

There is one thing that probably deserves a change to the code in `recursion-schemes`: the addition of the `TypeData` data constructor to `DatatypeVariant`. This causes an `-Wincomplete-patterns` warning when building `recursion-schemes` with `th-abstraction-0.5.*`:

```
[3 of 4] Compiling Data.Functor.Foldable.TH ( src/Data/Functor/Foldable/TH.hs, interpreted )

src/Data/Functor/Foldable/TH.hs:238:25: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In a case alternative:
        Patterns of type ‘DatatypeVariant’ not matched: TypeData
    |
238 |     isDataFamInstance = case variant of
    |                         ^^^^^^^^^^^^^^^...
```

`TypeData` represents `type data` declarations (introduced in GHC 9.6), and since `type data` data constructors do not exist at the term level, it doesn't make sense to define term-level instances for them. As such, I would propose to make it an error to use `recursion-schemes`' TH machinery with a `type data` declaration, just like it is currently an error to combine it with a data family instance.

--- a/recursion-schemes.cabal
+++ b/recursion-schemes.cabal
@@ -85,7 +85,7 @@ library
     build-depends:
       template-haskell >= 2.5.0.0 && < 2.20,
       base-orphans     >= 0.5.4   && < 0.9,
-      th-abstraction   >= 0.4     && < 0.5
+      th-abstraction   >= 0.4     && < 0.6
     exposed-modules:
       Data.Functor.Foldable.TH
 

--- a/src/Data/Functor/Foldable/TH.hs
+++ b/src/Data/Functor/Foldable/TH.hs
@@ -229,17 +229,22 @@ makePrimForDI rules mkInstance'
                 , datatypeInstTypes = instTys
                 , datatypeCons      = cons
                 , datatypeVariant   = variant }) = do
-    when isDataFamInstance $
-      fail "makeBaseFunctor: Data families are currently not supported."
+    checkAllowed
     makePrimForDI' rules mkInstance'
                    (variant == Newtype) tyName
                    (map toTyVarBndr instTys) cons
   where
-    isDataFamInstance = case variant of
-                          DataInstance    -> True
-                          NewtypeInstance -> True
-                          Datatype        -> False
-                          Newtype         -> False
+    checkAllowed =
+      case variant of
+        Datatype        -> pure ()
+        Newtype         -> pure ()
+        DataInstance    -> dataFamilyError
+        NewtypeInstance -> dataFamilyError
+#if MIN_VERSION_th_abstraction(0,5,0)
+        TH.Abs.TypeData -> fail "makeBaseFunctor: `type data` declarations are not supported."
+#endif
+
+    dataFamilyError = fail "makeBaseFunctor: Data families are currently not supported."
 
     toTyVarBndr :: Type -> TyVarBndrUnit
     toTyVarBndr (VarT n)          = plainTV n

